### DISCUSSION TITLE
Creating Change Summaries

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#2084] [Q&A] [answered]
Hi, when using ChangeSummaryManager.createChangeSummaries() to extract the changeset ids from the changesets in range index 0 to the latest, it starts downloading the changesets and successfully applies several of them to the temporary ChangeSummaryBriefcase.bim file. However, after several of the changesets have been downloaded, it ends up erroring out due to too many requests as shown below. It appears to be an API limit, but how do you use this function to create more than a few change summaries such that this doesn't happen? I've included the small snippet of code in which it occurs:

```
  private async getChangeSummaries(briefcase: BriefcaseDb) {
    const requestContext = await this.createContext();
    const changeSummaryIds = await ChangeSummaryManager.createChangeSummaries({ contextId: briefcase.contextId, iModelId: briefcase.iModelId, range: { first: 0, end: undefined }, requestContext });
    ChangeSummaryManager.attachChangeCache(briefcase);
    return changeSummaryIds;
  }
```

![image](https://user-images.githubusercontent.com/33571096/129869709-51542805-5118-4071-8cd5-2d54b0733673.png)

**Comment by pmconne** [comment, marked as answer]
Try upgrading your client as described in the answer to [this similar question](https://github.com/imodeljs/imodeljs/discussions/1491).

**Reply by nn15296** [reply]
@pmconne thanks for your response. I'm afraid I don't seem to have an upgrade button for the client?

**Reply by pmconne** [reply]
That's frustrating. @calebmshafer? @laurynasr?

**Reply by spowell-216** [reply]
@nn15296 this happened to me too, you will need to add at least on API at the bottom of the page, then you will get the upgrade button

**Reply by nn15296** [reply]
Still no upgrade button, but I decided to create a new client entirely. I was previously using the scopes in the screenshot below. These no longer seem to exist in the new app registration interface where "agent" has been replaced with "service", and I'm running into authorization issues. I've added the client as a team member and given it the same user roles as the previous client. Are you able to advise on which APIs and scopes relate to the scopes shown in the screenshot below, and whether I should be looking elsewhere to solve the authorization issues? Thanks for your help so far!

![image](https://user-images.githubusercontent.com/33571096/130470490-4f457e97-297d-43bd-b36e-19662c7ea5e4.png)

**Reply by aruniverse** [reply]
If you're on the latest imodel.js version, could you try the following before you initialize your AgentAuthClient:

```
Config.App.set("imjs_itwin_platform_authority", "https://ims.bentley.com")
```

**Reply by nn15296** [reply]
Having added that line of code, I'm getting the following error when I use the same scope as shown in my previous screenshot when initializing the AgentAuthClient. I think this is because the scope in the code doesn't match the scopes I assigned the client when I registered it:

![image](https://user-images.githubusercontent.com/33571096/130476960-e0f46cc4-ac03-44a5-a6ef-243135559a32.png)

When I do change the scope in the AgentAuthClient initialization to match those that I assigned when registering the client - where I assigned the scopes quite conservatively as just "itwinjs imodels:read", I end up with the following error:

![image](https://user-images.githubusercontent.com/33571096/130477153-1059614a-039f-4e55-ad13-42ce4e39e1df.png)

**Reply by nn15296** [reply]
@pmconne  I've managed to get the new client I just registered yesterday working using the itwinjs scope. Unfortunately, the rate limit issue I was having with the old client when trying to create multiple change summaries still seems to be present.

**Reply by pmconne** [reply]
I'm sorry to hear that. @calebmshafer, @ramanujam-raman, or @aruniverse can one of you assist, or suggest someone who can?

**Reply by calebmshafer** [reply]
@nn15296 a few questions just to try and figure out your workflow a bit better;

- Are you polling the iModel Hub for events using the iModelHub client [Events](https://www.itwinjs.org/learning/imodelhub/events/) API? 
- Are you handling a single iModel or multiple within the Agent?
- Last thing, would you be able to send the "x-correlation-id" associated with one of the failed requests? If you increase the logging level, it should show up in the failed requests details. You can increase the specific category by adding this line `Logger.setLevel(ITwinClientLoggerCategory.Request, LogLevel.Trace)`

**Reply by nn15296** [reply]
Hi @calebmshafer , thanks for the reply - I'll try and answer as best as I can to each point:

- I'm not using the Events API, I'm just extracting the change summaries from each changeset of a given iModel. I've tried using both the ChangeSummaryManager.createChangeSummaries(), and I've also tried manually walking the iModel forwards by going through each changeset for a given briefcase and updating the briefcase sequentially.
- I'm handling multiple iModels within the agent, but never more than one at any given time. The code below should give you an idea of what it's doing:
```
    // Get list of iModels from Project
    const ctx = await this.createContext();
    const hubIModelArray: HubIModel[] = await this.hubClient.iModels.get(ctx, this.config.CONTEXT_ID);
    ctx.enter();
    // Iterate through iModels
    for (const hubIModel of hubIModelArray) {
      if (hubIModel.id && hubIModel.name && hubIModelsToProcess.includes(hubIModel.name)) {
        const briefcaseProps: LocalBriefcaseProps = await this.getLocalBriefcaseProps(hubIModel.id);
        const changeSets: ChangeSet[] = (briefcaseProps?.changeSetId)
          ? await this.hubClient.changeSets.get(ctx, hubIModel.id, new ChangeSetQuery().fromId(briefcaseProps.changeSetId))
          : await this.hubClient.changeSets.get(ctx, hubIModel.id);
        ctx.enter();
        // Iterate through changesets
        for (const changeSet of changeSets) {
          if (changeSet.id) {
            console.log(`\n [${getTimeNow()}] Processing IModel: ${hubIModel.name} - Changeset: ${changeSet.id}`);

            // The updateLocalBriefcase function uses the BriefcaseManager.processChangesets() function to update the briefcase 
            // to the given changeSet id.
            await this.updateLocalBriefcase(briefcaseProps, changeSet.id);
            await this.processChangeset(briefcaseProps, hubIModel);

            // Delete ECChanges file for processed iModel to conserve space
            const ecChangesFile = join(this.cacheDirectory, "imodels", hubIModel.id, `${hubIModel.id}.bim.ecchanges`);
            if (existsSync(ecChangesFile)) {
              unlinkSync(ecChangesFile);
            }
            console.log(` [${getTimeNow()}] Finished Processing.`);
          }
        }
      }
    }
```

- I wasn't able to find the x-correlation-id I'm afraid. The output for the failed request, after increasing the logging level with the line of code you provided, was as follows:

```
Info    |itwin-client.Request| https://api.bentley.com/imodelhub/sv1.1/Repositories/iModel--b44166f7-1227-41dc-bf9b-fde9914fb1b0/iModelScope/ChangeSet/73ce5d690e69f24aa2a3c81d174670d3793906d0?
Trace   |itwin-client.Request| GET 200 https://api.bentley.com/imodelhub/sv1.1/Repositories/iModel--b44166f7-1227-41dc-bf9b-fde9914fb1b0/iModelScope/ChangeSet/73ce5d690e69f24aa2a3c81d174670d3793906d0 (276ms)
Info    |itwin-client.Request| https://api.bentley.com/imodelhub/sv1.1/Repositories/iModel--b44166f7-1227-41dc-bf9b-fde9914fb1b0/iModelScope/ChangeSet/?$filter=Index+eq+4&$top=1000
Trace   |itwin-client.Request| GET 429 https://api.bentley.com/imodelhub/sv1.1/Repositories/iModel--b44166f7-1227-41dc-bf9b-fde9914fb1b0/iModelScope/ChangeSet/?$filter=Index+eq+4&$top=1000 (108ms)
Error   |itwin-client.Request| 429 Error: Too Many Requests
```

**Reply by ramanujam-raman-zz** [reply]
@nn15296 , I'll reach out to you separately to find more details of your registration (to confirm the rate limits you are running into), and the code to figure how we can optimize it.

**Reply by nn15296** [reply]
@ramanujam-raman Sounds good. Thanks for your assistance!

**Reply by ramanujam-raman-zz** [reply]
@nn15296 Thanks for the registration and code details. Summarizing the diagnosis/fix for the benefit of others here. Please let us know if the fix below doesn't still resolve the issue. 

Quick Fix: Visit the app registration page - https://developer.bentley.com/my-apps/, pick the client, and click the  “Upgrade” button that should appear at the top right of the page. 

Diagnosis: We think the registration was first created without the “Visualization” scopes, and that caused it to use the default rate limits which are far lower than what we setup for Visualization (i.e., 50 per minute + 500 per hour instead of 500 per minute + 5000 per hour). That should have been automatically increased when you later updated the registration to add the Visualization scope – this is a bug at our end that we will fix. However, the above quick fix should take care of the problem in the meantime.

**Reply by nn15296** [reply]
@ramanujam-raman  Can confirm that the upgrade button has now appeared, and after upgrading, the rate limit issue is gone. Thanks everyone for your help!
